pandas/_typing.py

```diff
@@ -222,6 +233,7 @@ def __reversed__(self) -> Iterator[_T_co]:
 
 # to maintain type information across generic functions and parametrization
 T = TypeVar("T")
+S = TypeVar("S")
```

Comment:
IMHO, the use of `S` is confusing.  It's meaning is unclear when just looking at this file.  It's also only imported in `pandas/core/common.py`  So I'd prefer making `S` a `TypeVar` only within that file, and make it private (renaming to `_S`), and adding a comment that indicates it is needed because it is a second type within `pipe`

Reply:
(I would slightly prefer `_T` over `_S` but that's only my personal preference to signal that it basically is the same as `_typing.T`)

Reply:
See d4833701a2b73d148c074c9d91430b8c141038d7 for implementation of what's discussed here.